Plugins/Public/mobiledocking_plugin/Main.cpp

```diff
@@ -192,266 +89,707 @@ void LoadSettings()
-				if (pub::GetBaseID(mobiledockClients[client].proxyBaseID, (wstos(scProxyBase)).c_str()) == -4)
-				{
-					PrintUserCmdText(client, L"No proxy base in system detected. Contact a developer about this please. Required base: %ws", scProxyBase);
-					return;
-				}
+		}
 
-				mobiledockClients[client].carrierSystem = -1;
-			}
+		if (!shipAlreadyDocked)
+		{
+			PrintUserCmdText(dockingID, L"Carrier has no free docking capacity");
+			pub::Player::SendNNMessage(dockingID, pub::GetNicknameId("info_access_denied"));
 			return;
 		}
-		else if (carrier_client == -1)
-		{
-			// If the carrier doesn't exist for some reason, force the client to dock on it's last known base
-			uint iBaseID = mobiledockClients[client].iLastBaseID;
-			mobiledockClients[client].undockBase = Universe::get_base(iBaseID);
-			mobiledockClients[client].baseUndock = true;
+	}
 
-			return;
+	// Save the carrier info
+	auto carrierName = reinterpret_cast<const wchar_t*>(Players.GetActiveCharacterName(carrierID));
 
-		}
+	//In case this ship was launched from another mobiledock, unregister it first.
+	RemoveShipFromLists(dockingName, false);
 
-		//Get the carrier ship information
-		uint carrierShip;
-		pub::Player::GetShip(carrier_client, carrierShip);
+	if (!shipAlreadyDocked)
+	{
+		std::lock_guard<std::mutex> saveLock(saveMutex);
+
+		nameToCarrierInfoMap[carrierName].dockedShipList.emplace_back(dockingName);
+		nameToDockedInfoMap[dockingName].lastDockedSolar = Players[dockingID].iLastBaseID;
+		nameToDockedInfoMap[dockingName].carrierName = carrierName;
+		idToCarrierInfoMap[carrierID] = &nameToCarrierInfoMap[carrierName];
+		idToDockedInfoMap[dockingID] = &nameToDockedInfoMap[dockingName];
+		idToCarrierInfoMap[carrierID]->lastCarrierLogin = time(nullptr);
+		mobiledockClients[carrierID].iDockingModulesAvailable--;
+	}
 
-		uint clientShip;
-		pub::Player::GetShip(client, clientShip);
+	// Land the ship on the designated base
+	HkBeamById(dockingID, mobileDockingProxyBase);
+	PrintUserCmdText(carrierID, L"%ls successfully docked", dockingName);
+}
 
-		// Check to see if the carrier is currently in a base. If so, force the client to dock on that base.
-		if (!carrierShip)
+void HkTimerCheckKick()
+{
+	returncode = DEFAULT_RETURNCODE;
+
+	// DELAYEDDOCK vector iteration
+	for (auto& dockdata = dockingInProgress.begin(); dockdata != dockingInProgress.end();)
+	{
+		auto& dd = dockdata->second;
+		const auto& dockingShipID = Players[dd.dockingID].iShipID;
+		if (dockingShipID == 0)
+		{
+			dockdata = dockingInProgress.erase(dockdata);
+			continue;
+		}
+		Vector V1mov, V1rot;
+		pub::SpaceObj::GetMotion(dockingShipID, V1mov, V1rot);
+		if (V1mov.x > 5 || V1mov.y > 5 || V1mov.z > 5)
 		{
+			auto dockingName = reinterpret_cast<const wchar_t*>(Players.GetActiveCharacterName(dd.dockingID));
+			PrintUserCmdText(dd.dockingID, L"Docking aborted due to %ls's movement", dockingName);

```

Comment:
this isn't quite what I had in mind - only the carrier needs the dockingName sent to it